### DISCUSSION TITLE
Setting molecules as rigid bodies

### DISCUSSION THREAD

**Discussion by amruthesht** [glotzerlab/hoomd-blue#1779] [Support] [answered]
Hello All,

I have been running some simulations on a molecular system. I was wondering if there is an easy way to set all molecules as rigid bodies, where the central particles and moments are inherently calculated. I understand it can be done by explicitly following the procedure to build constituents from central particles. However, in my case, I have a particular system configuration I already have, which I want to turn rigid. Any help or tips on this are greatly appreciated.

Thank you!

Amru

**Comment by joaander** [comment, marked as answer]
This has been asked for many times over the years. I always encourage those interested to contribute a general code to the community that does this. For example, one could write a script that 
* Reads an input file where molecules are tagged.
* Writes a GSD file properly formatted with central and constituent particles.
* Also writes a json, pickle, or some other file type with the body type definitions for use by a HOOMD-blue script.

A general tool would:
* Discover all types of unique molecules in the input.
* For each type: Form the body definition in the local coordinate system.
* Diagonalize (if desired).
* Determine the final body positions and orientations.
* Create rigid body types and central particles.

Determining the unique bodies poses the most complex challenge. Due to numerical precision, what you consider to be the "same" molecule at different orientations in the box will appear to be many different molecules with slightly different local environments. This may be solvable with tolerances, but that requires much testing and debugging. There is also the problem of ordering. Are two molecules the same if the constituents are listed in a different order? You could work around these issues by requiring more explicit tag information in the input file - but then you are right back to the same level of information needed to define rigid bodies in the first place.

The only code I am aware of that attempts any of this is a [specific workflow](https://github.com/mosdef-hub/reproducibility_study/blob/main/reproducibility_project/hoomd4_subproject/src/engines/hoomd4/project.py) written with [mbuild](https://github.com/mosdef-hub/mbuild/). It constructs all molecules with the same particle ordering and the same **fixed orientation** to avoid the challenges I mentioned.

**Comment by amruthesht** [comment]
Thanks a lot for your detailed reply! If I go down the path of developing a detailed case of considering every molecule separately, I'll be happy to post it here!
